plutus-contract/src/Plutus/Contract/StateMachine.hs

```diff
@@ -331,10 +352,17 @@ runInitialise ::
     -- ^ The value locked by the contract at the beginning
     -> Contract w schema e state
 runInitialise StateMachineClient{scInstance} initialState initialValue = mapError (review _SMContractError) $ do
-    let StateMachineInstance{typedValidator, stateMachine} = scInstance
-        tx = mustPayToTheScript initialState (initialValue <> SM.threadTokenValue stateMachine)
-    let lookups = Constraints.typedValidatorLookups typedValidator
-    utx <- either (throwing _ConstraintResolutionError) pure (Constraints.mkTx lookups tx)
+    ownPK <- ownPubKey
+    utxo <- utxoAt (Ledger.pubKeyAddress ownPK)
+    let StateMachineInstance{stateMachine, typedValidator} = scInstance
+        red = Ledger.Redeemer (PlutusTx.toData (Scripts.validatorHash typedValidator, False))
+        constraints = mustPayToTheScript initialState (initialValue <> SM.threadTokenValue scInstance)
+            <> mustMintValueWithRedeemer red (SM.threadTokenValue scInstance)
+            <> maybe mempty (mustSpendPubKeyOutput . ttOutRef) (smThreadToken stateMachine)
```

Comment:
worth some comments

Reply:
In particular, it looks like there's a failure case here (where you get the mempty out of the `maybe` call) that there isn't in the line above. Can those come apart? Similarly when you destroy the token.

Reply:
I have tried to do this a bit, but constraints are not completely composable. For example, I tried splitting `mustPayToTheScript initialState (initialValue <> SM.threadTokenValue scInstance)` into separate `mustPayToTheScript initialState initialValue` and `mustPayToTheScript initialState (SM.threadTokenValue scInstance)` but that broke tests. (I didn't investigate further)

Reply:
I'm not saying it's wrong, I just want an argument explaining why it isn't wrong!

Reply:
Merging two `mustPayToTheScript` constraints is tricky (what if they have different datum values? What if the user actually wants to create two outputs?)

But I see that @sjoerdvisscher changed it to create the minting constraint and the must spend pubkey output constraint together, or not at all. So it's better now.